### DISCUSSION TITLE
How to forward option-enter (alt-enter) to terminal?

### DISCUSSION THREAD

**Discussion by schpet** [zed-industries/zed#18149] [Feature Requests]
i'm having trouble sending option-enter keypresses to the terminal, i've read this

https://zed.dev/docs/key-bindings#forward-keys-to-terminal

and in my keymap i've tried the following:

```json
[
  {
    "context": "Terminal",
    "bindings": {
      "alt-enter": ["terminal::SendKeystroke", "Alt+Enter"]
    }
  }
]
```

i've also tried the `option_as_meta` setting, and using the mac command key

```json
{
  "terminal": {
    "option_as_meta": true
  }
}
```

but no dice so far. i use option-enter to write newlines in [aider](https://aider.chat/).

**Comment by notpeter** [comment]
I believe it should be `"alt-enter"` not "Alt+Enter" (untested)

**Reply by schpet** [reply]
thanks for checking this out! i've tried these two combinations as well:

```
      "alt-enter": ["terminal::SendKeystroke", "alt-enter"]
      "alt-enter": ["terminal::SendKeystroke", "alt+enter"]
```

both with and without `option_as_meta` but no luck. i also observed that i cannot make this work with alacritty proper either, so maybe some sort of cursed meta/alt behaviour on mac os.

**Reply by schpet** [reply]
here's what i see without any alt-enter overrides in keymap.json, and `"option_as_meta": true`, zed on left, terminal.app on right

<img width="1114" alt="image" src="https://github.com/user-attachments/assets/4303b725-0bb8-41f8-98af-c996b30fc2db">

**Comment by notpeter** [comment]
Yeah, I just did the same thing but with fish_key_reader
<img width="439" alt="Screenshot 2024-09-25 at 10 30 29" src="https://github.com/user-attachments/assets/7c793227-7b95-43de-83a8-223ad8e80dd3">

We likely need to add `alt-enter` here:
https://github.com/zed-industries/zed/blob/cc7c0a31d97b578d7c6fd07011fd6d5cdca2d31b/crates/terminal/src/mappings/keys.rs#L52-L54

Separately if you have a buffer search deployed it is also broken. The keymap here could use tweaking
https://github.com/zed-industries/zed/blob/cc7c0a31d97b578d7c6fd07011fd6d5cdca2d31b/assets/keymaps/default-macos.json#L221-L229

**Reply by notpeter** [reply]
Thanks for reporting. Fixed in:
- https://github.com/zed-industries/zed/pull/18363

**Reply by schpet** [reply]
hey thanks so much for fixing this! also cool to learn about fish_key_reader, really straightforward way of understanding what's going on.
